niceman/resource/ssh.py

```diff
@@ -0,0 +1,132 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the niceman package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Environment sub-class to provide management of a SSH connection."""
+
+import attr
+import re
+from os import chmod
+from os.path import join
+from appdirs import AppDirs
+
+import logging
+lgr = logging.getLogger('niceman.resource.ssh')
+
+from .base import Resource, attrib
+import niceman.support.sshconnector2 # Needed for test patching to work.
+from ..ui import ui
+from ..support.starcluster.sshutils import SSHClient
+
+
+@attr.s
+class Ssh(Resource):
+
+    # Generic properties of any Resource
+    name = attr.ib()
+
+    # Configurable options for each "instance"
+    host = attrib(default=None,
+        doc="DNS or IP address of server.")
+    port = attrib(default=22,
+        doc="Port to connect to on remote host.")
+    key_filename = attrib(default=None,
+        doc="Path to SSH private key file matched with AWS key name parameter.") # SSH private key filename on local machine.
+    user = attrib(default=None,
+        doc="Username to use to log into remote environment.")
+    password = attrib(default=None,
+        doc="Password to use to log into remote environment.")
+
+    id = attr.ib(default=None)  # EC2 instance ID
+
+    type = attr.ib(default='ssh')  # Resource type
+
+    # Current instance properties, to be set by us, not augmented by user
+    status = attr.ib(default=None)
+
+    def connect(self):
+        """
+        Open a connection to the environment resource.
+        """
+        self._ssh = niceman.support.sshconnector2.SSHConnector2(self.host,
+            port=self.port,
+            key_filename=self.key_filename,
+            username=self.user,
+            password=self.password, # TODO: We can use password authentication, but how to securely save the password?
+        )
+
+    def create(self):
+        """
+        Create an EC2 instance.
+
+        Returns
+        -------
+        dict : config and state parameters to capture in the inventory file
+        """
+        self.id = self.host # TODO: The id should be a niceman-created unique identifier.
+        self.status = 'N/A'
+        return {
+            'id': self.id,
+            'status': self.status,
+            'host': self.host,
+            'user': self.user,
+            'port': self.port,
+            'key_filename': self.key_filename,
+        }
+
+    def delete(self):
+        return
+
+    def start(self):
+        return
+
+    def stop(self):
+        return
+
+    def execute_command(self, command, env=None):
```

Comment:
I think this now should be moved into `SSHSession(POSIXSession)` class under `_execute_command`, and `get_session` should be implemented here?